docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,242 @@
+
+```diff
+int ecall_with_user_check4(void* ptr) {
+
+    blob_t* blob;
+    blob_t volatile captured_blob;
+
+    // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Capture blob descriptors to avoid TOCTOU problems.
+    captured_blob.data = blob->data;
+    captured_blob.size = blob->size;
+
+    if (captured_blob.size > 4096)
+        return -1;
+
++   // Ensure validity of nested structure
++   if (!oe_is_outside_enclave(captured_blob.data, captured_blob.size))
++       return -1;
+
++   // Data parsed by render_data is "flat", non-self-referencing.
+    return render_data(captured_blob.data, captured_blob.size);
+}
+```
+## Our secure enclave function
+
+Thanks for sticking with us. Here is our corrected function:
+```c++
+int ecall_with_user_check5(void* ptr) {
+
+    blob_t* blob;
+    blob_t volatile captured_blob;
+
+    // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Capture blob descriptors to avoid TOCTOU problems.
+    captured_blob.data = blob->data;
+    captured_blob.size = blob->size;
+
+    if (captured_blob.size > 4096)
+        return -1;
+
+    // Ensure validity of nested structure.
+    if (!oe_is_outside_enclave(captured_blob.data, captured_blob.size))
+        return -1;
+
+    // Data parsed by render_data is "flat", non-self-referencing.
+    return render_data(captured_blob.data, captured_blob.size);
+}
+
+```
+
+<br />
+<br />
+<br />
+
+# Handling Secrets in Enclave Applications
+
+Open Enclave is an SDK that helps developers build apps that will run inside a hardware-based Trusted Execution Environment (TEE). At their core, TEEs protect application code and data at runtime from the host environment. Without new, hardware-implemented protections, a malicious or compromised host operating system would be able to modify code or read data. Any secrets managed by an application during runtime would be at risk of exposure. Open Enclave SDK helps developers build applications that are protected from these threats.
```

Comment:
Could be worth clarifying that Open Enclave does not provide code confidentiality.

Reply:
Good point Anand. I added:

> It's important to note that enclaves do not provide _code confidentiality_. This is noteworthy in the context of this discussion because, as you will read, hard-coded credentials in application code are to be avoided -- with or without enclave protections.